docs/development/etcdoperatortask-design.md

```diff
@@ -0,0 +1,501 @@
+    // The below block checks for the OperatorTask.
+    task := &v1alpha1.EtcdOperatorTask{}
+    if err := r.client.Get(ctx, req.NamespacedName, task); err != nil {
+        if client.IgnoreNotFound(err) != nil {
+            return reconcile.Result{}, err
+        }
+        return reconcile.Result{}, nil
+    }
+
+    logger := r.logger.WithValues("runId", string(controller.ReconcileIDFromContext(ctx)))
+    // create a instance of TaskHandler
+    taskHandlerInstance := r.createTaskHandlerInstance(task)
+    
+    // Triggers the deletion flow in case the task is in a completed state or if it has been marked for deletion.
+    if task.IsCompleted() || task.IsMarkedForDeletion() {
+        return r.triggerDeletionFlow(ctx, taskHandlerInstance, task).ReconcileResult()
+    }
+    // triggers the task execution flow.
+    return r.reconcileTask(ctx, client.ObjectKeyFromObject(task), taskHandlerInstance).ReconcileResult()
+}
+
+```
+
+```go
+// reconcileTask manages finalizer, execution, and status updates.
+func (r *Reconciler) reconcileTask(ctx context.Context, taskObjKey client.ObjectKey, taskHandler tasks.TaskHandler) ctrlutils.ReconcileStepResult {
+    reconcileStepFns := []reconcileFn{
+        r.recordTaskStartOperation,
+        r.ensureTaskFinalizer,
+        r.updateStatusObservedGeneration,
+        r.transitionToPendingState, // updates the status field. Skipped if already done.
+        r.admitTaskPreconditions, // calls the implemented interface method. Run only at the first time.
+        r.transitionToInProgressState, // updates the status. Skipped if already marked.
+        r.runTask, // calls the implemented interface method.
+    }
+
+    for _, step := range reconcileStepFns {
+        ctx.Logger.Info("Executing step", "step", step)
+        result := step(ctx, taskObjKey, taskHandler)
+        if ctrlutils.ShortCircuitReconcileFlow(result) {
+            return result
+        }
+    }
+
+    return ctrlutils.ReconcileAfter(task.Spec.TTLSecondsAfterFinished, "Task completed, waiting for TTL to expire")
+}
+```
+</details>
+
+---
+
+
+### Deletion Flow
+
+<details>
+<summary>Show Task Deletion Flow</summary>
+
+```go
+func (r *Reconciler) triggerTaskDeletionFlow(
+    ctx context.Context,
+    logger logr.Logger,
+    taskObjKey client.ObjectKey,
+    taskHandler tasks.TaskHandler,
+) ctrlutils.ReconcileStepResult {
+    if task.IsCompleted() && !task.IsMarkedForDeletion() {
+        if !task.TTLHasExpired() {
+            return ctrlutils.ReconcileAfter(task.Spec.TTLSecondsAfterFinished, "Task completed, waiting for TTL to expire")
+        }
+    }
+
+    deletionStepFns := []reconcileFn{
+        r.recordTaskDeletionStartOperation, // update the status. Skip if already done.
+        r.cleanupTaskResources, // Cleanup of any kind of resources used to run the task. Runs the interface method.
+        r.recordTaskDeletionSuccessOperation,
```

Comment:
Might not be necessary, since the resource is about to be deleted anyway. We can save an additional status update call here.

Reply:
Removed